compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs

```diff
@@ -2177,7 +2177,13 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
                 }
             }
             Res::Def(DefKind::AssocTy, def_id) => {
-                debug_assert!(path.segments.len() >= 2);
+                if path.segments.len() < 2 {
+                    let guar = self
+                        .dcx()
+                        .struct_span_err(span, "cannot infer type, type annotations needed")
```

Comment:
This might seem minor but the phrasing "cannot infer type" isn't quite 'right' in this context since HIR ty lowering doesn't actually use type inference (in the `rustc_infer` sense) or trait solving in ItemCtxts (roughly "contexts where `_` is not permitted") when resolving associated types to projections, hence the phrasing "**ambiguous** associated type/constant" as used by `probe_single_bound…`